### ISSUE TITLE
object:get_pos() returns nil in entity on_step

### ISSUE THREAD

**Issue by corarona** [minetest/minetest#12656]
##### Minetest version
5.6 release

##### OS / Hardware
Operating system: Debian testing
CPU: i5-2540M  @ 2.60GHz
GPU model :Intel 2nd Generation Core Processor Family Integrated Graphics Controller (rev 09)
OpenGL version:  3.3 

##### Summary
Not exactly sure how (yet) but an entity's get_pos() can be made to return nil in on_step.

##### Steps to reproduce
1. install MineClone2 and start a game
2. grant yourself give privilge: /grantme give
3. give yourself a bow (/giveme mcl_bows:bow) and some arrows (/giveme mcl_bows:arrow -1)
4. Enchant it with the "flame" enchantment: while wielding the bow /forceenchant singleplayer(or your playername) flame
5. Shoot a bunch of arrows (it takes like 5-20 for me usually)
6. observe the game crash:

```
2022-08-04 03:43:12: ERROR[Main]: ServerError: AsyncErr: Lua: Runtime error from mod '??' in callback luaentity_Step(): Invalid position (expected table got nil).
2022-08-04 03:43:12: ERROR[Main]: stack traceback:
2022-08-04 03:43:12: ERROR[Main]: 	[C]: in function 'get_objects_inside_radius'
2022-08-04 03:43:12: ERROR[Main]: 	...st/bin/../games/MineClone2/mods/ITEMS/mcl_bows/arrow.lua:148: in function 'on_step_old'
2022-08-04 03:43:12: ERROR[Main]: 	...mes/MineClone2/mods/ENVIRONMENT/mcl_void_damage/init.lua:24: in function <...mes/MineClone2/mods/ENVIRONMENT/mcl_void_damage/init.lua:23>
```

I'm sure we just did everything wrong and this is expected behavior in which case feel free to close and it's easy enough to fix in lua too. However this does not crash in 5.5.1 so I thought i'd let you know (if it wasn't a crash i would not bother you).

references:
https://git.minetest.land/MineClone2/MineClone2/issues/2517
https://git.minetest.land/MineClone2/MineClone2/pulls/2518

**Comment by sfan5** [comment]
`ARROW_ENTITY.on_step` (line 117) -> `mcl_burning.tick` -> `mcl_util.deal_damage` -> `set_hp` (may kill object)
`ARROW_ENTITY.on_step` (line 121) -> `get_pos` (returns nil)

This changed in 303329f2d6bc96f7756c72845de1bd87f62174a1 but is your bug for not tracking object deletion.

Also I think you should have verified the hypothesis by putting a `get_pos` call on the very first line of on_step, you'd have found that the pos only becomes nil later.

**Comment by corarona** [comment]
totally non breaking change i suppose ^^

**Comment by sfan5** [comment]
By a strict definition bugfixes are indeed breaking changes.

**Comment by corarona** [comment]
Note how i literally called your reaction in the issue hehe.

look I understand your time is extremely valuable but I don't think you're doing yourself or the project  a favor by being so dismissive every single time (I know this is not just you personally but you seem to be a great driver of minetests toxicity). 

I know several other people feel the same way too so don't even try to say I'm just too sensitive or whatever.

I'm not going to bother you in the future unless it's absolutely necessary - and even then chances seem good it will be pointless.

Minetest could be so nice without mods or users huh?

**Comment by sfan5** [comment]
Can you help me understand and explain where and in what way I was being dismissive in this issue? Because I honestly don't get it.
The closest I can see is my remark that you could potentially have figured out the bug yourself. After writing my response I contemplated that it wasn't so necessary to say this but editing it out afterwards felt stupid too.

**Comment by corarona** [comment]
The dismissive part is implementing such a "bug fix" for which you could easily have anticipated that it would introduce huge problems (and i kind of doubt this will only be problematic in mcl2). And then acting like everyone else is stupid and immediately close the issue.

You do realize that every game needs to look at every single on_step now and make sure such a thing can't happen ?

I mean couldn't the bugfix have let the step complete and then delete the object ? seems to me like the more intuitive way, particularly because it would not introduce new crashes and more closely resemble the (non buggy part of) the behavior before.

Knowing this mcl2 probably won't be able to support 5.6 for a while - and honestly the fact that such things can just happen kind of makes me think ultimately nothing from the minetest api can be relied upon (at least in entity on_step).

I just don't get why you need to break things every single release (and at least from what I understand somewhat needlessly so at least in this case).

Funny enough the one breaking change everyone wants (larger worlds, what we currently have barely deserves the name world honestly) is constantly being refused even with prs available that would actually be somewhat backwards compatible.

**Comment by sfan5** [comment]
Sorry what? These are all technical disagreements.

You went so far as to accuse me of being a "great driver of toxicity", dismissed talking to me as pointless (without mentioning any reasons, those only came after I asked), made some vague hints that I am driving modders and users away from Minetest (???).
Your entire answer was basically written to be intentionally hurtful and convey "You are bad, mean and not worth talking to".
Do you consider this a constructive communication style? Why didn't you make a honest attempt to convey your arguments to me?
I realize communicating tone over the internet is hard but I really think I have shown goodwill here that was only met with harsh responses.


We can gladly talk about the technical side -- I assure you we don't change things without reason -- but it'd be nice to hear some sort of apology from you.

**Comment by corarona** [comment]
Well you're driving them away from communicating here yes. Because every time issues are just closed as "fix it yourself" which apart from being fairly impolite just makes you think "ok i can spend my time more productively".

And no I do not consider this a constructive communication style. But as I implied i did not expect much constructive communication here. I just realized that I will be busy fixing the fallout of this for the next weeks probably which admittedly made me angry.

And I mean from mcl pov its just after the png behavior shuffling around between 5.3, 5.4 and 5.5 - we're using a lua tga encoder for the maps now so it doesn't break again btw. I mean i did not personally "suffer" from that one too much since others jumped to do the work on it hehe but i still remember the drama well.

Now saying that you personally are a driver of toxicity was presumptuous I really do not know that maybe i should have made that clearer (i did say "seem to be though").

Yes I shouldn't have said some of that stuff. I'm sorry.

As for the technical side I think this was a pretty reckless move and I doubt mcl will be the only thing having issues with this. It was clearly never even considered and now there does not seem to be any desire to address it.

I think this is a lot more breaking than you imply (and maybe honestly think).

But yes I suppose I can spend my time more productively :P

**Comment by corarona** [comment]
it was pointed out to me it was the dynamic media api not png stuff. As I said i did not work on that.

**Comment by sfan5** [comment]
Fair enough, apology accepted.

> a "bug fix" for which you could easily have anticipated that it would introduce huge problems (and i kind of doubt this will only be problematic in mcl2)

To be honest I didn't expect any regressions when writing that fix. In hindsight it's easier to tell that there are mods (accidentally or intentionally) relying on dead entities not being removed

> couldn't the bugfix have let the step complete and then delete the object ? seems to me like the more intuitive way, particularly because it would not introduce new crashes and more closely resemble the (non buggy part of) the behavior before.

It could but I think it's not ideal to introduce delays where they are not technically necessary.
And would `on_death` run on the same step or the next? There's potential for confusion here.

Not sure what you mean by the behaviour before. The old behaviour was that entities that have been set to zero HP would never disappear. Changing that is bound to break mod code that doesn't expect this, even if you delay it one step.

> the fact that such things can just happen kind of makes me think ultimately nothing from the minetest api can be relied upon

You can definitely rely on the things documented in lua_api.txt.
Not in the strict definition of "if it once worked it always works" but more like "if I use the API as intended and don't rely on buggy behaviour then it will continue to work". lua_api.txt is unfortunately underspecified in this area and especially about how certain callbacks are ordered.

That an entity is removed when you set HP to zero (= dead) makes perfect sense to me and fwiw the bug reporter had the exact same expectation (#11975).

> Because every time issues are just closed as "fix it yourself" which apart from being fairly impolite

If it's about issues being closed with the first response then I can see why that isn't the nicest experience, however fundamentally I don't think that would have changed anything. Afaik other devs agree that with the fix and the "delay it one step" solution doesn't have good chances of being accepted.

> [...] the png behavior shuffling around between 5.3, 5.4 and 5.5 [...]

I think this is another thing that I'm unfairly being blamed for.
Yes, the API made an u-turn in how it works (and not because it was convenient, the old way was not viable to keep). But also it had a proper deprecation period.

5.3 code worked on 5.4 but raised a warning, if you then corrected your code it would work both on 5.4 and future 5.5. Writing code that works on 5.3, 5.4 and 5.5 is a bit more involved but possible.
The issue erlehmann opened about this (if you remember) was complete nonsense because it was just "the engine forced me to provide a callback so I put an empty one, it doesn't work anymore!".
And yes, these are all breaking changes under a strict definition, but I don't see practical usefulness in that (see #12004 for that).

> I just realized that I will be busy fixing the fallout of this for the next weeks probably [...]. I think this is a lot more breaking than you imply (and maybe honestly think).

For MCL you can work around this easily.
IIRC you already have a callback that wraps all entity on_step's (voiddamage was it?), then in `deal_damage` instead of `set_hp(0)` you set an object property to defer the killing to the end of the step.

Returning early from the burning damage would still be preferable for correctness reasons but if you have workaround in place you can work on that anytime.
